torch/csrc/jit/script/compiler.cpp

```diff
@@ -569,15 +568,40 @@ struct to_ir {
   }
 
   void runCleanupPasses(std::shared_ptr<Graph>& to_clean) {
+    // NB: moving returns to end needs to run before all other passes
+    moveAllReturnsToEnd(to_clean);
     // remove any uses of tuples that we inserted that are not needed
     LowerSimpleTuples(to_clean);
     ConstantPooling(to_clean);
   }
 
-  FunctionSchema emitDef(
-      const Def& def,
-      const Self& self,
-      Block* block) {
+  // If the graph might not return, at an implicit None return at the end
+  void handleMaybeNoReturn(const Def& def, Block* block) {
+    // if we're emitting a block that isn't the graph block,
+    // (like in a closure), do not add a return
+    if (block != block->owningGraph()->block()) {
+      return;
+    }
+
+    if (exit_blocks.count(graph->block()) == 0) {
+      auto decl_ret = def_stack_.back().declared_return_type_;
+      if (decl_ret && decl_ret != NoneType::get() &&
```

Comment:
If someone intents to return an Optional[T], I don't think they are ever going to simply omit the `return None`. I'd rather be more strict and just say that a function returning a value is missing a return statement.